l1/l1.go

```diff
@@ -75,101 +62,86 @@ func (c *Client) checkChainID(ctx context.Context) error {
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
-		case err := <-subHeaders.Err():
-			c.log.Warnw("L1 header subscription failed, resubscribing", "error", err)
-			subHeaders.Unsubscribe()
+		case <-ticker.C:
+			latestL1BlockNumber, finalisedL1BlockNumber, latestL1BlockHash, err := c.l1.LatestAndFinalisedHeights(ctx)
+			if err != nil {
+				c.log.Warnw("Failed to fetch finalised and latest L1 block numbers", "err", err)
+				continue
+			}
 
-			subHeaders, err = c.subscribeToHeaders(ctx, headerChan)
+			if c.finalisedHeight > finalisedL1BlockNumber {
+				// TODO how should we handle such catastrophic events?
+				return errReorgFinalisedBlock
+			}
+
+			c.finalisedHeight = finalisedL1BlockNumber
+
+			head, err := c.l1.StarknetHead(ctx, latestL1BlockHash)
 			if err != nil {
-				return err
+				c.log.Warnw("Failed to fetch Starknet L1 head", "err", err)
+				continue
 			}
-			defer subHeaders.Unsubscribe() //nolint:gocritic
-		case header := <-headerChan:
-			l1Height := header.Number.Uint64()
-			c.log.Debugw("Received L1 header", "number", l1Height, "hash", header.Hash().Hex())
-		Outer:
-			// Check for updates.
-			// We need to loop in case there were multiple LogStateUpdates emitted.
-			for {
-				select {
-				case err := <-subUpdates.Err():
-					c.log.Warnw("L1 update subscription failed, resubscribing", "error", err)
-					subUpdates.Unsubscribe()
-
-					subUpdates, err = c.subscribeToUpdates(ctx, logStateUpdateChan)
-					if err != nil {
-						return err
+			c.log.Debugw("Retrieved L1 head", "latestL1", latestL1BlockNumber, "finalisedL1",
+				finalisedL1BlockNumber, "l2BlockNumber", head.BlockNumber)
+
+			c.nonFinalisedQueue[latestL1BlockNumber] = head
+			var maxFinalisedNumber uint64
+			for l1BlockNumber := range c.nonFinalisedQueue {
+				if l1BlockNumber > latestL1BlockNumber {
+					// Ethereum block was reorged. Remove l1Head from queue.
+					delete(c.nonFinalisedQueue, l1BlockNumber)
+				} else if l1BlockNumber <= finalisedL1BlockNumber && l1BlockNumber >= maxFinalisedNumber {
+					if l1BlockNumber >= maxFinalisedNumber {
+						maxFinalisedNumber = l1BlockNumber
 					}
-					defer subUpdates.Unsubscribe() //nolint:gocritic
-				case logStateUpdate := <-logStateUpdateChan:
-					c.log.Debugw("Received L1 LogStateUpdate",
-						"number", logStateUpdate.BlockNumber,
-						"stateRoot", logStateUpdate.GlobalRoot.Text(felt.Base16),
-						"blockHash", logStateUpdate.BlockHash.Text(felt.Base16))
-					if logStateUpdate.Raw.Removed {
-						// NOTE: we only modify the local confirmationQueue upon receiving reorged logs.
-						// We assume new logs will soon follow, so we don't notify the l2Chain.
-						c.confirmationQueue.Reorg(logStateUpdate.Raw.BlockNumber)
-					} else {
-						c.confirmationQueue.Enqueue(logStateUpdate)
-					}
-				default:
-					break Outer
 				}
 			}
 
-			if err := c.setL1Height(l1Height); err != nil {
+			maxFinalisedHead := c.nonFinalisedQueue[maxFinalisedNumber]
```

Comment:
I think it is better to use `x, ok := someMap[someKey]`, where ok is either `true` or `false` depending on whether `someKey` exists or not.